File: nodetodo/src/controller/todoController.ts
import {
  TAG_TODO_CREATE,
  TAG_TODO_DELETE,
  TAG_TODO_EDIT,
} from "../constants/AppConstant";
import { TodoRepository } from "../db/todoRepository";
import { isDeleteFeatureEnable, logEvent } from "../util/util";

const todoRepository = new TodoRepository();

/**
 * A controller class to handle
 * Create,Delete,Update,GetById,GetAll operation
 * through api's
 */
export class TodoController {
  async create(req: any, res: any) {
    try {
      const todo = await todoRepository.create(req.body);
      res.json(todo);
      logEvent(TAG_TODO_CREATE, "TODO created", todo);
    } catch (error) {
      console.log(error);
      res.status(500).json({ error: "Failed to create todo" });
    }
  }

  async update(req: any, res: any) {
    try {
      const data = await todoRepository.update(req.body);
      res.json(data);
      logEvent(TAG_TODO_EDIT, "TODO edited", data);
    } catch (error) {
      console.log(error);
      res.status(500).json({ error: "Failed to update todo" });
    }
  }

  async getById(req: any, res: any) {
    try {
      const todo = await todoRepository.getById(req.params.id);
      res.json(todo);
    } catch (error) {
      res.status(404).json({ error: "Todo not found" });
    }
  }

  async deleteById(req: any, res: any) {
    const enableDelete = await isDeleteFeatureEnable();

    if (!enableDelete) {
      res.status(401).json({ error: "Not authorized to delete todo" });
    } else {
      try {
        const id = await todoRepository.deleteById(req.params.id);
        logEvent(TAG_TODO_DELETE, "TODO deleted", { id: id });
        res.json({ id: id });
      } catch (error) {
        console.log(error);
        res.status(404).json({ error: "Failed to delete Id" });
      }
    }
  }

  async getAll(req: any, res: any) {
    try {
      const todos = await todoRepository.getAll();
      res.json(todos);
    } catch (error) {
      console.log(error);
      res.status(500).json({ error: "Failed to fetch todos" });
    }
  }
}

File: golangtodo/src/todocontroller.go
package src

import (
	"encoding/json"
	"fmt"
	"net/http"
	"strconv"

	"github.com/gorilla/mux"
)

type TodoController struct {
	repo *TodoRepository
}

func NewTodoController(repo *TodoRepository) *TodoController {
	return &TodoController{repo: repo}
}

func (c *TodoController) GetAllTodos(w http.ResponseWriter, r *http.Request) {

	todos, err := c.repo.GetAll()
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	json.NewEncoder(w).Encode(todos)
}

func (c *TodoController) GetTodoByID(w http.ResponseWriter, r *http.Request) {

	params := mux.Vars(r)
	id, err := strconv.ParseUint(params["id"], 10, 64)
	fmt.Println(params)
	if err != nil {
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}

	todo, err := c.repo.GetByID(uint(id))
	if err != nil {
		http.Error(w, err.Error(), http.StatusNotFound)
		return
	}

	json.NewEncoder(w).Encode(todo)
}

func (c *TodoController) CreateTodo(w http.ResponseWriter, r *http.Request) {
	var newTodo Todo
	err := json.NewDecoder(r.Body).Decode(&newTodo)
	if err != nil {
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}

	todo, err := c.repo.Create(&newTodo)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	w.WriteHeader(http.StatusCreated)
	json.NewEncoder(w).Encode(todo)
	mapTodo := structToMap(*todo)
	LogEvent(TAG_TODO_CREATE, mapTodo)

}

func (c *TodoController) UpdateTodo(w http.ResponseWriter, r *http.Request) {
	var qtodo Todo
	err := json.NewDecoder(r.Body).Decode(&qtodo)

	todo, err := c.repo.GetByID(qtodo.ID)
	if err != nil {
		http.Error(w, err.Error(), http.StatusNotFound)
		return
	}

	err = c.repo.Update(&qtodo)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	w.WriteHeader(http.StatusOK)
	json.NewEncoder(w).Encode(qtodo)

	mapTodo := structToMap(*todo)
	LogEvent(TAG_TODO_EDIT, mapTodo)

}

func (c *TodoController) DeleteTodo(w http.ResponseWriter, r *http.Request) {

	isDeleteEnable := IsDeleteFeatureEnabled()

	if isDeleteEnable {
		params := mux.Vars(r)
		id, err := strconv.ParseUint(params["id"], 10, 64)
		fmt.Println(params)
		if err != nil {
			http.Error(w, err.Error(), http.StatusBadRequest)
			return
		}

		err = c.repo.Delete(uint(id))
		if err != nil {
			http.Error(w, err.Error(), http.StatusInternalServerError)
			return
		}
		w.WriteHeader(http.StatusOK)
		json.NewEncoder(w).Encode(map[string]int{"id": int(id)})
		LogEvent(TAG_TODO_DELETE, map[string]string{"id": params["id"]})

	} else {
		w.WriteHeader(http.StatusNonAuthoritativeInfo)
		json.NewEncoder(w).Encode(map[string]string{"error": "Not authorized to delete todo"})
	}

}
